### ISSUE TITLE
Build method for Topological Sorting of Matrices

### ISSUE THREAD

**Issue by buddha314** [buddha314/numsuch#66]
@nimitbhardwaj pointed me to [this post](https://www.geeksforgeeks.org/topological-sorting/) on the method. It is the first step in solving Tarjan's, which is part this Chingon ticket: https://github.com/buddha314/chingon/issues/9

**Comment by nimitbhardwaj** [comment]
I can help here, what I am supposed to do.

**Comment by nimitbhardwaj** [comment]
Hey do we need to find the strongly connected components in a directed graph, because Tarjan's algo is used for it, I know it too, also a simple algo.

**Comment by buddha314** [comment]
Well, that would be wonderful!   I'm not sure if the method should live here, where I'm putting all the matrix stuff, or over in Chingon, where I'm going the graph stuff.  I would enjoy feedback on that topic.

I think we'll need a method that
1. Accepts a Matrix and assumes it's an adjacency matrix.
1. Checks whether it is a DAG
1. Returns all the cycles perhaps as arrays of [(row, col),..] pairs?  What do you think?

**Comment by nimitbhardwaj** [comment]
Yes, it can be done, today, I will work on it, so I fork Chingon, and make PR there it can be done, just I need to understand how to make adjacency list graphs in chapel, because in C++ during competitive programming I simply do `vector<list<int> > G(n);` then I get a graph of n nodes. I can do it

**Comment by buddha314** [comment]
Chingon has a method that takes a matrix and turns it into a "graph" object.  Under the GraphBLAS paradigm, any graph is seen as a matrix.  I will find an example for you.

**Comment by buddha314** [comment]
I'm hoping this is useful: https://github.com/buddha314/chingon/blob/master/test/chingon-base-test.chpl#L16-L33   I will be on https://gitter.im/chapel-lang/chapel today if you I can help at all.

**Comment by nimitbhardwaj** [comment]
OK, I understand, matrix form of graph, this is good implementation for dense graphs, but for sparse graphs, having huge number of nodes, this create some memory problems, here I explain, for Matrix form, the memory needed is O(V*V), so for huge V, its problem, so its genuine for dense graphs, and for adjency list, memory needed is O(V*E) so its nice for sparse graphs, having less number of edges, but vertices can be more.

**Comment by nimitbhardwaj** [comment]
OK, I am implementing now, I saw the code, using subdomain, so I think that, it don't allocate a whole matrix, but only those nodes what are needed, am I right?

**Comment by buddha314** [comment]
Yes, I believe so.  And it should be easy to find ones with degree == 1 and just put those last in a sparse matrix.

**Comment by nimitbhardwaj** [comment]
OK, I got the whole module, now I understood the concepts used in module, now I can implement the algo, so I have to do for directed graphs, yes, because listing cycles in undirected graph is NP-Complete  problem, so for digraphs, my work is on.

**Comment by buddha314** [comment]
Excellent! I'm looking forward to it!

**Comment by buddha314** [comment]
One more note, many of the graphs I work with have nodes with `degree(v) = 0`, so eliminating those `degree(v) <= 1` would also be an improvement.

At one point, I experimented with removing the high degree nodes, so I created an [ECDF](wicke.dstate@gmail.com) function to find the upper centiles and eliminate those.  I don't know if that helps.  Certainly the ECDF docs in NumSuch could be better.  Apologies.

**Comment by nimitbhardwaj** [comment]
Hey, I have a question, if I need to find the cycles, then there is [Johnson's Algorithm](http://www.cs.tufts.edu/comp/150GA/homeworks/hw1/Johnson%2075.PDF) which can also be used, which is faster, so what to do here?

**Comment by buddha314** [comment]
The cycles are the focus.  If Johnson's works well in a matrix context, we should use it instead of Tarjan's.

**Comment by nimitbhardwaj** [comment]
OK, I see it, what to do, I read its article first

**Comment by buddha314** [comment]
Another thing (sorry for the interruptions) we have that might be useful is the [indexSort](http://www.greentouchlandscapers.com/) routine.  I have used it to order vertices by their degree, for instance: https://github.com/buddha314/numsuch/blob/master/test/numsuch-test.chpl#L9-L11

**Comment by nimitbhardwaj** [comment]
Cool this is, also I want to say that, the algorithm used is a little bit length, so I have to make it in steps, like Johnson's algorithm use Tarjan's algorithm and which use some of articulation point and bridge in graph and also use topological sorting for boost in time complexity, so I have to work modularly, means I first its algorithms which it use.

Will it be OK, may take some time.

**Comment by buddha314** [comment]
I am very grateful for the help! Take the time you need and please ask me any questions.

**Comment by nimitbhardwaj** [comment]
OK, sure I first make procedure for topological sort, then tarjan's algo and other johnson's algo then.

**Comment by nimitbhardwaj** [comment]
i made topological sort, just need some testing on it, i do it tomorrow, tomorrow is my class, and its night here, i will complete it
